Tickets/dm 10050

mxk62 opened the pull request.

mxk62 committed:
Adapt DAX generator the most recent GraphML specs

Modified DAX generator to support following attributes:
- for file nodes: 'urls' and 'sites',
- for task nodes: 'streams'.

mxk62 committed:
Fix various errors

There were a few trivial errors related to misusing APIs of the
third-party libraries used by the generator.  Also, due to recent
changes in Butler, Pathogen stopped returning correct paths to files.  I
made necessary changes to get everything back to the working condition.

mxk62 committed:
Merge _check_sanity with _color

The auxiliary function called _check_sanity was checking if the graph is
bipartite.  As this is a prerequisite for graph coloring to identify
files and tasks, it is more reasonable to move that test there rather
that keep it separate.

mxk62 committed:
Improve DAX initialization

DAX generator can be initialized by passing a preexisting graph
representing the workflow.  However, it could lead to unexpected side
effects if graph would by modified afterwards as graphs are intrnally
represented as dictionaries.  I modified the DAX generator to make it
use the graph's copy instead to prevent theses situations from
happening.

mxk62 committed:
Enhance format recognition

While reading a file with a persisted workflow, DAX generator selects
appropriate import method based on the file's extension.  Added few more
variants representing the supported formats.

mxk62 committed:
Remove redundant coloring

Removed redundant graph coloring from write() method.  Now graph is
colored when initialized or read from an external file.

mxk62 committed:
Add testsuite for DAX generator

mgckind commented on daxgen/daxgen.py:
```diff
@@ -16,8 +17,8 @@ class Daxgen(object):
-        self.graph = graph if graph is not None else nx.DiGraph()
-        self._check_sanity()
+        self.graph = graph.copy() if graph is not None else nx.DiGraph()
+        self._color()
```
I think the name of this private method might be confusing to what the method is doing, especially is called a couple of times within other methods

mxk62 replied:
I'll change it to `_label()`.

mgckind replied:
👍

mgckind commented on daxgen/daxgen.py:
```diff
@@ -42,14 +43,16 @@ def read(self, filename):
         except KeyError:
             raise ValueError('Format \'%s\' is not supported yet.' % ext)
-        self._check_sanity()
+        self._color()
```
see above

mgckind commented on daxgen/daxgen.py:
```diff
@@ -42,14 +43,16 @@ def read(self, filename):
         }
         ext = filename.split('.')[-1]
         try:
             self.graph = methods[ext.lower()](filename)
```
what if the reader itself raises an error?

mxk62 replied:
It would mean that file does not exists or is malformed. I consider such a situation to be a critical (unrecoverable) error and just let a reader raise an exception whatever it feels is appropriate. Do you think it should be handled differently?

mgckind replied:
at first I thought the Exception raised by the reader would be hidden behind this try, and the message would be lost but it would raise anyways as this only captures the KeyError. I think this is fine,

mgckind commented on daxgen/daxgen.py:
```diff
@@ -42,14 +43,16 @@ def read(self, filename):
         try:
             self.graph = methods[ext.lower()](filename)
         except KeyError:
             raise ValueError('Format \'%s\' is not supported yet.' % ext)
```
This might be clearer:
`"Format '{0}' is not supported yet".format(ext)`

mgckind commented on daxgen/daxgen.py:
```diff
@@ -1,9 +1,10 @@
-_FILE_ATTRS = {'lfn', 'pfns'}
+# Set of mandatory attributes unique to file nodes. Do not have to be
+# exhaustive.
+_FILE_ATTRS = {'lfn'}
```
what does lfn stand for?

mxk62 replied:
`lfn` stands for _logical file name_ and, similarly, `pfn` stands for _physical file name_. _Pegasus_ documentation uses this shortcuts routinely and I wasn't able to come up with something equally concise but more meaningful.

mgckind replied:
it makes sense, probably you can add those acronyms as commented lines?

mgckind commented on daxgen/daxgen.py:
```diff
@@ -77,19 +79,35 @@ def write(self, filename, name='dax'):
+                sites = attrs.get('sites')
+                if sites is None:
+                    sites = ','.join(len(urls) * ['local'])
+                for url, site in zip(urls.split(','), sites.split(',')):
```
are `sites` and `urls` forced to be the same length?

mxk62 replied:
Correct, each `url` has to have a corresponding `site`. I think _Pegasus_ would do it for me (I mean it would use `local` be default) but I prefer to do it in my code in case we would like to use different default settings.

mgckind replied:
👍

mgckind commented on daxgen/daxgen.py:
```diff
@@ -77,19 +79,35 @@ def write(self, filename, name='dax'):
         # Add files to DAX-level replica catalog.
         catalog = {}
         for v in files:
```
probably file_id would be better than v, same with task_id below.

mxk62 replied:
I _reluctantly_ agree. ;)

mgckind commented on daxgen/daxgen.py:
```diff
@@ -77,19 +79,35 @@ def write(self, filename, name='dax'):
+            # respective Pegasus file object.
+            args = attrs['args'].split()
+            lfns = list(set(catalog) & set(args))
+            if lfns:
```
what happens if this is empty?

mxk62 replied:
What is empty, `args` or `lfns`? If `lfns` is empty it means that there is nothing to do with the job/task arguments and that's fine.

If you worried about `args`, I guess you're right. I'll have to add the safeguard which will catch possible `KeyError` if the node doesn't have that attribute.

mgckind replied:
I meant `lfns` and later `args` as well. not sure when args might be empty but it might be worth to check for it

mgckind commented on daxgen/daxgen.py:
```diff
@@ -105,23 +123,19 @@ def write(self, filename, name='dax'):
-                    elif stream == 'stdout':
+                streams = attrs.get('streams')
+                if streams is not None:
+                    if streams & 1 != 0:
```
is the same as `if streams == 1: ` right?

mxk62 replied:
Not exactly. Both `stderr` and `stdout` could be redirected to the same file and then `streams` would be set to 3. So by checking which bits are set I'm able to tell exactly what streams are redirected to the file in question.

mgckind replied:
I see, got it.

mgckind commented on daxgen/daxgen.py:
```diff
@@ -164,6 +161,20 @@ def _check_sanity(self):
+            # V represents tasks, or the other way round.
+            v = next(iter(U))
+            node_attrs = set(self.graph.node[v].keys())
+            files, tasks = (U, V) if node_attrs >= _FILE_ATTRS else (V, U)
```
is this the only use of _FILE_ATTRS? I mean you used that attribute before 'lfn' in other parts of the code.

Why not just `if _FILE_ATTRS.issubset(node_attrs)` ?

mxk62 replied:
Yes, `_FILE_ATTRS` is used only here. Do you think I should move its definition to he `_color()` method?

And agree, your version of checking the condition shows my intentions in a clearer way. I'll change the code accordingly.

mgckind replied:
yes, maybe move it to the method itself especially if meant to be private and not a constant through the code, unless you want to add other attributes later on. Or you can use it every time you check for this particular attribute in other parts of the code

mgckind commented on daxgen/tests/test_daxgen.py:
```diff
@@ -0,0 +1,122 @@
+        for u, v in zip(sorted(g.nodes()), sorted(h.nodes())):
+            self.assertEqual(g.node[u], h.node[v])
+
+    def test_write(self):
```
is this skipped for now? or just a placeholder for an upcoming test?

mxk62 replied:
Correct, it is just a placeholder at the moment. I'm planning to extend the testsuite in the next cycle.

mgckind replied:
👍

mgckind commented on daxgen/tests/test_daxgen.py:
```diff
@@ -0,0 +1,122 @@
+        attrs = tuple(g.node[u]['bipartite'] for u in ['0', '1', '2'])
+        self.assertEqual(attrs, (1, 0, 1))
+
+    def test_read_unknown(self):
```
unknown file or format?

mxk62 replied:
Yes, the test is supposed to check code's behaviour if a file with unknown format is used. So obviously I need to change it. How about `test_read_unsupported`?

mgckind replied:
it'd be clearer to see what the test does by the name only, especially when running the unittests

mxk62 committed:
Make various improvements

Addressing reviewers suggestions, I made several minor modifications
improving code readability and robustness.
